### PR TITLE
Add support SUNION command

### PR DESCRIPTION
**Pull request by argsno**
Add Support for the `SUNION` Command.

Descriptions:
Returns the members of the set resulting from the union of all the given sets.
Keys that do not exist are considered to be empty sets.

refer: https://github.com/microsoft/garnet/issues/181

And will support `SUNIONSTORE` soon.

### CODE REVIEW

**Comment by TalZaccai** [comment]
@argsno marked this as draft for now, please mark it as "ready for review" when you're done :)

File: libs/server/Objects/Set/SetObject.cs [hunk 1]
```diff
@@ -192,5 +193,7 @@ public override unsafe void Scan(long start, out List<byte[]> items, out long cu
             if (cursor == set.Count)
                 cursor = 0;
         }
+
+        public HashSet<byte[]> Set => set;
```
Comment (TalZaccai) [inline libs/server/Objects/Set/SetObject.cs:197]: The Set object should not be made public

File: libs/server/Storage/Session/ObjectStore/SetOps.cs [hunk 2]
```diff
@@ -361,6 +362,33 @@ public unsafe GarnetStatus SetScan<TObjectContext>(ArgSlice key, long cursor, st
...
+        public GarnetStatus SetUnion<TObjectContext>(ArgSlice[] keys, out HashSet<byte[]> output, ref TObjectContext objectStoreContext)
+            where TObjectContext : ITsavoriteContext<byte[], IGarnetObject, SpanByte, GarnetObjectStoreOutput, long>
+        {
+            output = new HashSet<byte[]>(new ByteArrayComparer());
+
+            foreach (var key in keys)
+            {
+                if (GET(key.ToArray(), out var currObject, ref objectStoreContext) == GarnetStatus.OK)
+                {
+                    var currSet = ((SetObject)currObject.garnetObject).Set;
```
Comment (TalZaccai) [inline libs/server/Storage/Session/ObjectStore/SetOps.cs:383]: You don't interact directly with the set object, you need to go through the storage layer to do that. Please refer to other commands that interact with more than one object (for example: LMOVE, ZDIFF)

File: libs/server/Resp/RespCommand.cs [hunk 3]
```diff
@@ -693,6 +693,10 @@ static RespCommand MatchedNone(RespServerSession session, int oldReadHead)
                                         {
                                             return (RespCommand.STRLEN, 0);
                                         }
+                                        else if (*(ulong*)(ptr + 2) == MemoryMarshal.Read<ulong>("\r\nSUNION\r\n"u8))
```
Comment (TalZaccai) [inline libs/server/Resp/RespCommand.cs:696]: Should be `*(ulong*)(ptr + 4) == MemoryMarshal.Read<ulong>("SUNION\r\n"u8)`

File: libs/server/Resp/Objects/SetCommands.cs [hunk 4]
```diff
@@ -94,6 +94,62 @@ private unsafe bool SetAdd<TGarnetApi>(int count, byte* ptr, ref TGarnetApi stor
...
+        /// </summary>
+        /// <param name="count"></param>
+        /// <param name="ptr"></param>
+        /// <param name="storageApi"></param>
+        /// <typeparam name="TGarnetApi"></typeparam>
+        /// <returns></returns>
+        private bool SetUnion<TGarnetApi>(int count, byte* ptr, ref TGarnetApi storageApi)
+            where TGarnetApi : IGarnetApi
+        {
+            if (count < 2)
```
Comment (TalZaccai) [inline libs/server/Resp/Objects/SetCommands.cs:109]: I think count == 1 is also supported, it just returns the members of the set

File: libs/server/Resp/Objects/SetCommands.cs [hunk 5]
```diff
@@ -94,6 +94,62 @@ private unsafe bool SetAdd<TGarnetApi>(int count, byte* ptr, ref TGarnetApi stor
...
+            }
+
+            storageApi.SetUnion(keys, out var result);
+
+            // write the size of result
+            var resultCount = result?.Count ?? 0;
+            while (!RespWriteUtils.WriteArrayLength(resultCount, ref dcurr, dend))
+                SendAndReset();
+
+            if (result != null)
```
Comment (TalZaccai) [inline libs/server/Resp/Objects/SetCommands.cs:139]: In what case would result be null?

**Review by TalZaccai — CHANGES_REQUESTED** [review]

**Review by TalZaccai — COMMENTED** [review]
Hey @argsno! Any updates on this PR? Are you planning on adding SUNIONSTORE to it? LMK :) Thanks!

**Comment by argsno** [comment]
> Hey @argsno! Any updates on this PR? Are you planning on adding SUNIONSTORE to it? LMK :) Thanks!

@TalZaccai Sorry, I've been a little busy lately.  There is no time to support SUNIONSTORE, I will fix the current comments as soon as possible

**Comment by TalZaccai** [comment]
> > Hey @argsno! Any updates on this PR? Are you planning on adding SUNIONSTORE to it? LMK :) Thanks!
> 
> @TalZaccai Sorry, I've been a little busy lately. There is no time to support SUNIONSTORE, I will fix the current comments as soon as possible

Ah bummer! No worries, I'll reopen the work item for someone else to pick up... Thanks again!

**Review by TalZaccai — APPROVED** [review]

**Pull request merged**